design/local-node-agent.md

```diff
@@ -91,17 +107,72 @@ The experience is also normalized across other distributed storage platforms tha
+For CRD create operations, an error will be returned if a CRD with that name already exists, and a duplicate CRD will not be created.
+Furthermore, for update operations, an error will be returned if an attempt to update an existing CRD occurs that specifies an out of date version of the object.
+
+##### **ReadWriteOnce**
+For `ReadWriteOnce`, the agent needs to ensure that only one client is accessing the volume at any time.
+During the `Mount()` operation, the agent will look for an existing CRD by its primary key, the PVC ID.
+If no CRD currently exists, then the agent will create one, signifying that it has won exclusive access on the volume, then proceed with the attach and mount.
+The CRD that the agent creates will contain the pod, node and mount dir of the current attachment.
+
+If a CRD does already exist, the agent will check its existing attachments list.
+If the list specifies that the volume is attached to a different pod than the one we are currently mounting for, then another consumer already has exclusive access of the volume and the agent must honor `ReadWriteOnce` and fail the operation.
+However, if the previous attachment is for the **same** pod and namespace that we are currently mounting for, this means that the volume is being failed over to a new node and was not properly cleaned up on its previous node.
+Therefore, the agent will "break" the old lock by removing the old attachment entry from the list and adding itself, then continuing with attaching and mounting as usual.
+
+##### **ReadWriteMany**
+For `ReadWriteMany`, the agent will allow multiple entries in the attachments list of the CRD.  When `Mount()` is called, the agent will either create a new CRD instance if it does not already exist, or simply add a new attachment entry for itself to the existing CRD.
+
+##### **Detach**
+During the detach operation, the CRD will be deleted to signify that the volume has been unlocked (or updated to remove the entry from the attachment list of the CRD for the case of `ReadWriteMany`).
+
+However, if the node where the volume was attached dies, the agent on that node may not get a chance to perform the detach and update the CRD instance.
+In this case, we will need to clean up the stale record and potentially perform the detach later on if the node returns.
+This will be performed in two separate ways:
+1. As previously mentioned, the agent will check for existing attachments when it is requested to `Attach()`.  If it finds that the attachment belongs to a node that no longer exists, it will "break" the lock as described above.
+1. Periodically (once per day), the operator will scan all volume attachment CRD instances, looking for any that belong to a node that no longer exists.
+
+In both cases, when a "stale" attachment record is found, its details will be added to a volume attachment garbage collection list, indexed by node name.
```

Comment:
An use case is when a stale record becomes a good record. ie, node dies and volume is attached to a different node. A 'stale' record is added to the 'node cleaner'. Then the original node comes back and the pod moves back to the original node. There should be a way to 'remove' this stale record.

Reply:
I'm not sure if i understand that scenario entirely, because I would not expect Kubernetes to immediately schedule the pod back to its original node as soon as the node comes back.  I would expect the pod to just stay on the 2nd node (the one it failed over to).  I'll try to clarify the wording here that the agent is responsible for doing everything it can to clean up stale records. though.